syntax = "proto3";

package mindtest;

enum MindTestType {
    MBTI = 0;
    ENNEAGRAM = 1;
    BIG_FIVE = 2;
}

enum QuestionType {
    SINGLE_CHOICE = 0;
    MULTI_CHOICE = 1;
    TRUE_FALSE = 2;
}

enum MBTIType {
    INTROVERT = 0;
    EXTROVERT = 1;
    SENSING = 2;
    INTUITION = 3;
    THINKING = 4;
    FEELING = 5;
    JUDGING = 6;
    PERCEIVING = 7;
}

enum EnneagramType {
    TYPE_1 = 0;
    TYPE_2 = 1;
    TYPE_3 = 2;
    TYPE_4 = 3;
    TYPE_5 = 4;
    TYPE_6 = 5;
    TYPE_7 = 6;
    TYPE_8 = 7;
    TYPE_9 = 8;
}

enum BigFiveType {
    EXTRAVERSION = 0;
    AGREEABLENESS = 1;
    CONSCIENTIOUSNESS = 2;
    NEUROTICISM = 3;
    OPENNESS = 4;
}

message Question {
    int32 id = 1;
    string text = 2;
    QuestionType type = 3;
    repeated string options = 4;
}

message MindTest {
    int32 id = 1;
    MindTestType type = 2;
    repeated Question questions = 3;
}

message Answer {
    int32 question_id = 1;
    string selected_option = 2;
}

message WebToServerRequest {
    int32 test_id = 1;
}

message WebToServerResponse {
    MindTest test = 1;
}

message ServerToWebRequest {
    int32 test_id = 1;
    repeated Answer answers = 2;
}

message ServerToWebResponse {
    oneof message {
        MBTIResultResponse mbti_result = 1;
        EnneagramResultResponse enneagram_result = 2;
        BigFiveResultResponse big_five_result = 3;
    }
}

message MBTIResultResponse {
    MBTIType personality_type = 1;
    string description = 2;
}

message EnneagramResultResponse {
    EnneagramType enneagram_type = 1;
    string description = 2;
}

message BigFiveResultResponse {
    BigFiveType big_five_type = 1;
    string description = 2;
}

service Server {
    rpc web_to_server(WebToServerRequest) returns (WebToServerResponse);
    rpc server_to_web(ServerToWebRequest) returns (ServerToWebResponse);
}